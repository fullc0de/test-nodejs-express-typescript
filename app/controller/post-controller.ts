import {HttpRequest, HttpResponse} from "./interfaces";

export class PostController {

    public async index(request: HttpRequest): Promise<HttpResponse> {
        return {
            headers: { "Hello": "World" },
            statusCode: 200,
            body: { message: "wow success!" }
        }
    }

    public async show(request: HttpRequest): Promise<HttpResponse> {
        const params = request.params
        return {
            headers: { "Hello": "World" },
            statusCode: 200,
            body: {
                id: params.id,
                author: "J. K. Rolling",
                title: "Harry Potter"
            }
        }
    }
}
