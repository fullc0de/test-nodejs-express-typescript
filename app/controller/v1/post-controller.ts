import {Context} from "../common-interfaces";
import BaseController from "../base-controller";
import ControllerInterface from "../interface/controller-interface";
import { Route } from "../../decorator/route";

@Route('posts')
export class PostController extends BaseController implements ControllerInterface {
    
    public async index(ctx: Context) {

        ctx.response = {
            statusCode: 200,
            body: { message: "wow success! V1" }
        };
    }

    public async show(ctx: Context) {
        const postId = super.validateParamId(ctx.request.params.id)
        if (postId instanceof Error) {
            throw postId
        }
        
        ctx.response = {
            statusCode: 200,
            body: {
                id: postId,
                author: "J. K. Rolling",
                title: "Harry Potter"
            }
        };
    }
}
