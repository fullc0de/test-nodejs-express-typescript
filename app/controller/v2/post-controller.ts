import { Context, Route, UserAuth, PostParam, QueryParam } from "deco-versioning-router";
import { PostController as V1 } from "../v1/post-controller";
import { JwtAuthDecoInjector } from '../../deco-injector/jwt-auth-deco-injector';
import { Users } from '../../model/users';
import validator from 'validator';

@Route("posts", "v2")
@UserAuth(new JwtAuthDecoInjector())
export class PostController extends V1 {

    @QueryParam("testParam", { required: true })
    public async index(ctx: Context) {
        const user: Users = ctx.additional['user'];

        ctx.response = {
            headers: { "Hello": "World" },
            statusCode: 200,
            body: { 
                message: "wow success! V2",
                author: `${user.firstName} ${user.lastName}`
            }
        };
    }

    @PostParam("hello", { required: false })
    @PostParam("age", { required: true, validate: (v) => validator.isInt(v), errorMessage: "'age' param must be an integer." })
    public async post(ctx: Context) {
        ctx.response = {
            statusCode: 200,
            body: { 
                message: "posting done V2",
                dummy: ctx.request.query.dummyQuery,
                age: ctx.request.body.age,
                hello: ctx.request.body.hello
            }
        };
    }
}
