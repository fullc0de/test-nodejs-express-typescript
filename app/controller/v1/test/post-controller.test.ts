import { PostController } from "../post-controller"

describe('PostController V1', () => {
    let controller = new PostController();
    describe('- check validation: post_id', () => {
        it('should return true', () => {
            expect(controller.validateParamId('123'))
            .toEqual(123);
        });
        it('should return false', () => {
            expect(controller.validateParamId('Ab2'))
            .toEqual(new Error('invalid resource parameter (Ab2)'));
        });
    });
});