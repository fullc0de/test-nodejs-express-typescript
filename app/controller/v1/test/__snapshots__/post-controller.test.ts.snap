// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostController V1 - check functionality of base controller get current user 1`] = `
Users {
  "address": "verve",
  "createdAt": 2020-01-14T21:48:50.862Z,
  "firstName": "hwang",
  "id": 1,
  "lastName": "kyokook",
  "updatedAt": 2020-01-14T21:48:50.862Z,
  "version": 1,
}
`;
